### ISSUE TITLE
Bug: `ESLint` class should not load plugins already defined in `plugins`

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#15079]
### Environment

Node version: v16.9.1
npm version: v7.9.0
Local ESLint version: v7.32.0
Global ESLint version: Not found
Operating System: win32 10.0.19043

### What parser are you using?

Default (Espree)

### What did you do?

```js
import {ESLint} from 'eslint'

new ESLint({
	baseConfig: {
		extends: [
			'plugin:foo/preset',
		],
	},
	plugins: {
		foo: {
			configs: {
				preset: {}
			}
		}
	}
})
```

### What did you expect to happen?

No errors.

### What actually happened?

The `config-array-factory` is trying to load the plugin that I've set in the `plugins`.

```js
node:internal/modules/cjs/loader:933
  const err = new Error(message);
              ^

Error: Failed to load plugin 'foo' declared in 'BaseConfig': Cannot find module 'eslint-plugin-foo'
Require stack:
- E:\project\@forks\eslint-plugin-unicorn\__placeholder__.js
Referenced from: BaseConfig
	// ...
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

N/A

**Comment by nzakas** [comment]
Strange this is only popping up now. Did this used to happen with CLIEngine?

**Comment by fisker** [comment]
I don't know, never used `plugins` before.

**Comment by mdjermanovic** [comment]
CLIEngine didn't have a constructor option that accepts plugin implementations, but an instance method [`CLIEngine#addPlugin`](https://eslint.org/docs/developer-guide/nodejs-api#cliengineaddplugin). As building a config array for `baseConfig` happens right away during `new CLIEngine()`, I believe this functionality didn't exist - it wasn't possible to define `plugin:foo/preset` in `baseConfig` and provide the `foo` plugin as an object instead of loading it from the file system.

**Comment by mdjermanovic** [comment]
> If we add `plugins` to the [`CLIEngine` constructor](https://github.com/eslint/eslint/blob/e3cd1414489ceda460d593ac7e7b14f8ad45d4fc/lib/cli-engine/cli-engine.js#L574), we can init [`additionalPluginPool`](https://github.com/eslint/eslint/blob/e3cd1414489ceda460d593ac7e7b14f8ad45d4fc/lib/cli-engine/cli-engine.js#L586) with the plugin implementations, will it solve the problem?

I think something like that would work. In the current version, we are updating `additionalPluginPool` on the instance, but that's too late for `extends` in `baseConfig`.

https://github.com/eslint/eslint/blob/e3cd1414489ceda460d593ac7e7b14f8ad45d4fc/lib/eslint/eslint.js#L435-L454

We could instead pass `additionalPluginPool` map to the constructor. CLIEngine is now an internal module, so new options on it wouldn't be public.

**Comment by nzakas** [comment]
We shouldn’t be making any further changes to CLIEngine at this point, as it will be going away. We need to start transitioning functionality into the ESLint class directly.

**Comment by nzakas** [comment]
I took a look at this and updating CLIEngine is probably the easiest way to fix this.
